packages/plugin-core/src/components/lookup/QuickPickTemplateSelector.ts

```diff
@@ -0,0 +1,101 @@
+import {
+  ConfigUtils,
+  DendronError,
+  DLogger,
+  NoteProps,
+} from "@dendronhq/common-all";
+import { ExtensionProvider } from "../../ExtensionProvider";
+import { NoteLookupProviderUtils } from "./NoteLookupProviderUtils";
+import { HistoryEvent } from "@dendronhq/engine-server";
+import { Logger } from "../../logger";
+import * as vscode from "vscode";
+import { Disposable } from "vscode";
+import { VSCodeUtils } from "../../vsCodeUtils";
+import { DendronContext } from "../../constants";
+import { AutoCompletableRegistrar } from "../../utils/registers/AutoCompletableRegistrar";
+import { AutoCompleter } from "../../utils/autoCompleter";
+
+export class QuickPickTemplateSelector {
+  getTemplate(opts: {
+    logger?: DLogger;
+    providerId?: string;
+  }): Promise<NoteProps | undefined> {
+    const logger = opts.logger || Logger;
+    const id = opts.providerId || "TemplateSelector";
+    const extension = ExtensionProvider.getExtension();
+    const controller = extension.lookupControllerFactory.create({
+      nodeType: "note",
+      buttons: [],
+    });
+    const provider = extension.noteLookupProviderFactory.create(id, {
+      allowNewNote: false,
+      forceAsIsPickerValueUsage: true,
+    });
+    const config = extension.getDWorkspace().config;
+
+    const tempPrefix = ConfigUtils.getCommands(config).templateHierarchy;
+    const initialValue = tempPrefix ? `${tempPrefix}.` : undefined;
+
+    let disposable: Disposable;
+    return new Promise((resolve) => {
+      NoteLookupProviderUtils.subscribe({
+        id,
+        controller,
+        logger,
+        onHide: () => {
+          resolve(undefined);
+          NoteLookupProviderUtils.cleanup({
+            id,
+            controller,
+          });
+        },
+        onDone: (event: HistoryEvent) => {
+          const data = event.data;
+          if (data.cancel) {
+            resolve(undefined);
+          } else {
+            const templateNote = event.data.selectedItems[0] as NoteProps;
+            resolve(templateNote);
+          }
+          NoteLookupProviderUtils.cleanup({
+            id,
+            controller,
+          });
+          disposable?.dispose();
+          VSCodeUtils.setContext(DendronContext.NOTE_LOOK_UP_ACTIVE, false);
+        },
+        onError: (event: HistoryEvent) => {
+          const error = event.data.error as DendronError;
+          vscode.window.showErrorMessage(error.message);
+          resolve(undefined);
+          NoteLookupProviderUtils.cleanup({
+            id,
+            controller,
+          });
+          disposable?.dispose();
+          VSCodeUtils.setContext(DendronContext.NOTE_LOOK_UP_ACTIVE, false);
+        },
+      });
+      controller.show({
+        title: "Select template to apply",
+        placeholder: "template",
+        provider,
+        initialValue,
+      });
+
+      VSCodeUtils.setContext(DendronContext.NOTE_LOOK_UP_ACTIVE, true);
```

Comment:
@jonathanyeung 

Doing some cleanup here after a rebase to apply the changes from https://github.com/dendronhq/dendron/pull/3559/ to work here as well.

This code as-is works if `QuickPickTemplateSelector` is used in the context of another command (i.e. `ApplyTemplateCommand`), but not if this is part of `NoteLookupCommand`.

This is because as part of `onDidHide`, the lookup controller sets `DendronContext.NOTE_LOOKUP_ACTIVE` to false, and that timing happens to be after the quickpick for template selection is shown. This results in the template selector not autocompletable.

Since we are relying on the global context state, it's a bit tricky to handle this if another lookup has to be chained. Any pointers on how to tackle this?

Reply:
One possible 'solution' is to remove the conditional check on tab autocomplete of having the note-lookup-active flag set.  This _theoretically_ shouldn't be a problem, b/c it uses a pub/sub event, and if there are no lookup controls around when someone is hitting tab, then no callback gets invoked.

The potential concern with that approach is I'm not 100% confident all of our lookup controller cleanup is working properly.  I don't think the side effects should be too bad if some controller _is_ still hanging around? Maybe something we can test w/ nightly. 

Maybe this is a defeatist mentality haha, but I also wouldn't think it's so bad if tab autocomplete doesn't work in this one control - it's still an improvement from where we are today.

Another reason to refactor the lookup code...

Reply:
> This theoretically shouldn't be a problem, b/c it uses a pub/sub event, and if there are no lookup controls around when someone is hitting tab, then no callback gets invoked.

But we are contributing a default keybinding for the autocomplete command (as `tab`) as an extension, and without that context key would we be blocking tabs completely?

Reply:
Resolved by removing the `onDidHide` behavior of the quickpick and handling it during command cleanup